workflow-manager/task/task.go

```diff
@@ -0,0 +1,222 @@
+// package task contains representations of tasks that are sent to facilitator
+// workers by workflow-manager. A _task_ is a work item at the data share
+// processor application layer (i.e., intake of a batch or aggregation of many
+// batches); its name is chosen to distinguish from Kubernetes-level _jobs_.
+package task
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"sync"
+	"time"
+
+	"github.com/letsencrypt/prio-server/workflow-manager/utils"
+
+	"cloud.google.com/go/pubsub"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/sns"
+)
+
+// Task is a task that can be enqueued into an Enqueuer
+type Task interface {
+	// MarshalToJson marshals self into a JSON message
+	MarshalToJson() ([]byte, error)
+}
+
+// Aggregation represents an aggregation task
+type Aggregation struct {
+	// AggregationID is the identifier for the aggregation
+	AggregationID string `json:"aggregation-id"`
+	// AggregationStart is the start of the range of time covered by the
+	// aggregation in UTC, with minute precision, formatted like
+	// "2006/01/02/15/04"
+	AggregationStart string `json:"aggregation-start"`
+	// AggregationEnd is the end of the range of time covered by the aggregation
+	// in UTC, with minute precision, formatted like "2006/01/02/15/04"
+	AggregationEnd string `json:"aggregation-end"`
+	// Batches is the list of batch ID date pairs of the batches aggregated by
+	// this task
+	Batches []Batch `json:"batches"`
+}
+
+func (a Aggregation) MarshalToJson() ([]byte, error) {
+	return json.Marshal(a)
+}
+
+// Batch represents a batch included in an aggregation task
+type Batch struct {
+	// ID is the batch ID. Typically a UUID.
+	ID string `json:"id"`
+	// Time is the timestamp on the batch, in UTC, with minute precision,
+	// formatted like "2006/01/02/15/04"
```

Comment:
Since we have specific requirements for the timestamp format, this would be a nice thing to enforce at the API boundary between the `task` package and `main`. For instance, instead of letting `main` directly instantiate `Batch`, you could provide a function that builds `Batches`. That function would take a `*time.Time` (Go's internal time type), and format it in the expected wire format.

Alternately, you could make the field of type `SlashyMinuteTime` (a type alias for `*time.Time), and define a custom JSON marshaller for `SlashyMinuteTime`. That would let you continue instantiating `Batch` directly, while also enforcing correct wire format.

Reply:
Go type aliases are a bit of a pain because you have to type assert back to the original thing to use any of its methods, but I still like the idea. I introduced `type Timestamp time.Time` with a custom JSON formatter. I'm curious though: is there a reason for it to be an alias to `*time.Time` instead of just `time.Time`?

Reply:
> Go type aliases are a bit of a pain because you have to type assert back to the original thing to use any of its methods, but I still like the idea.

For type aliases, you're actually doing a type conversion, not a type assertion, which is both safer and syntactically somewhat cleaner:

https://www.sohamkamani.com/golang/type-assertions-vs-type-conversions/

Short version: a type assertion is fallible and done at runtime. A type conversion is infallible and checked at compile time.

> is there a reason for it to be an alias to *time.Time instead of just time.Time?

Nope. I can never remember whether it's typically passed as a pointer or not, so I took a guess.